Add generate-terraform-data command

iomodo opened the pull request:
#### Summary
This PR adds `generate-terraform-data` command

#### Ticket Link
  Fixes https://mattermost.atlassian.net/browse/MM-33535

catalintomai commented on README.md:
```diff
@@ -57,6 +57,8 @@ Only once you need to run `go run ./cmd/appsctl/ provision bucket` to create the
 
 To provision an app run `go run ./cmd/appsctl/ provision app /PATH/TO/YOUR/APP/BUNDLE`. Use `--update` to update the lambda functions if they already exist. This option should not be used in production.
 
+Run `go run ./cmd/appsctl/ generate-terraform-data /PATH/TO/YOUR/APP/BUNDLE` to generate terraform variables. It will print the indented json directly in the console. Pipe as needed.
```
We should make clear who is to invoke this script and in what context (e.g. the cloud team pipeline, etc.). That will also give info on decisions like to `createAWSClient**WithoutCredentials**)

catalintomai commented on server/api/impl/aws/aws_keys.go:
```diff
@@ -14,6 +14,13 @@ import (
+	lambdaFunctionFileNameMaxSize = 64
+	appIDLengthLimit              = 32
+	versionFormat                 = "v00.00.000"
+	awsStaticAssetsPrefix         = "static"
```
nit: do we need the `aws` prefix (we are in the AWS package/namespace)

catalintomai commented on server/api/impl/aws/provision_data.go:
```diff
@@ -0,0 +1,175 @@
+		if err != nil {
+			return nil, errors.Wrap(err, "can't get function name")
+		}
+		functionMap[function.Name] = FunctionData{
```
Is this the (TerraForm) standard way of packaging these AWS assets/artifacts for pipeline consumption - or is it an internal contract between us and the cloud team? If it's the first option, can we reference some TF documentation for future reference?

catalintomai commented on server/api/impl/aws/provision_data.go:
```diff
@@ -0,0 +1,175 @@
+
+const bundleStaticAssetsFolder = "static/"
+
+type ProvisionData struct {
```
Is there anything here TerraForrm specific or is it an internal data model shared with the cloud team?

iomodo replied:
Nothing terraform specific just an internal model

hanzei replied:
Should we document it somewhere?

catalintomai reviewed (approved):
few comments, LGTM.

levb reviewed (approved):
I am not presently set up to test this, but "on paper" looks good.

hanzei reviewed (changes requested):
Should we document the outputted data?

iomodo commented on server/api/impl/aws/provision_app.go:
```diff
@@ -197,25 +101,20 @@ func isValid(url string) bool {
 }
 
-func (c *Client) provisionFunctions(manifest *apps.Manifest, functions []functionInstallData, shouldUpdate bool) error {
+func (c *Client) provisionFunctions(manifest *apps.Manifest, functions map[string]FunctionData, shouldUpdate bool) error {
```
we already have the map generating a slice and passing it to `provisionFunctions` would be additional unnecessary work.

iomodo replied:
Map is already there converting to the slice will be extra work

iomodo commented on server/api/impl/aws/provision_data.go:
```diff
@@ -0,0 +1,175 @@
+	S3BucketName    string                  `json:"s3_bucket_name"`
+	StaticFiles     map[string]AssetData    `json:"static_files"`
+	LambdaFunctions map[string]FunctionData `json:"lambda_functions"`
+	Manifest        *apps.Manifest          `json:"-"`
```
We store a pointer to `Manifest` everywhere. It's just for consistency.

iomodo commented on server/api/impl/aws/provision_data.go:
```diff
@@ -0,0 +1,175 @@
+}
+
+type FunctionData struct {
+	Bundle  io.Reader `json:"-"`
```
Good question. On server we are using `json:"-"` extensively that's why I chose it here, but I'm all for on unexporting the fields.

iomodo commented on server/api/impl/aws/provision_data.go:
```diff
@@ -0,0 +1,175 @@
+type ProvisionData struct {
+	S3BucketName    string                  `json:"s3_bucket_name"`
+	StaticFiles     map[string]AssetData    `json:"static_files"`
+	LambdaFunctions map[string]FunctionData `json:"lambda_functions"`
```
key is the file name in the bundle, field is the function name that should be provisioned

hanzei replied:
It isn't obvious that the key is the file name. Does it make sense to add a `FileName` field an covert the map to a slice?

iomodo replied:
file name is the name of the `zip` file where function bundle is. We need `map` for mapping filenames and Function data. I think we could just document it.

hanzei commented on server/utils/utils.go:
```diff
@@ -2,6 +2,9 @@ package utils
 	"encoding/json"
+	"os"
+
+	"github.com/mattermost/mattermost-server/utils/fileutils"
```
```suggestion
	"github.com/mattermost/mattermost-server/v5/utils/fileutils"
```

hanzei commented on apps/app.go:
```diff
@@ -45,11 +45,11 @@ func (v AppVersion) IsValid() error {
 		}
 
-		if c == '-' || c == '_' {
+		if c == '-' || c == '_' || c == '.' {
```
We shouldn't forbid dots in versions as it's commonly used. `getFunctionName` does sanitize any dots.

iomodo replied:
Developer can have the version "v1.0.0" in the manifest. We will sanitize it later. This method is used to check if the manifest is valid or not. Am I missing something?

hanzei reviewed (commented):
When I try to run `$ go run ./cmd/appsctl/ provision app  $PATH_TO_BUNDLE`

it panics with

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0xc38bfc]

goroutine 1 [running]:
github.com/mattermost/mattermost-plugin-apps/server/api/impl/aws.(*Client).S3FileUpload(0xc00025d540, 0xc000120300, 0x2b, 0x7f72a43d5208, 0xc0001440f0, 0xc000120300, 0x2b)
	/home/sumacheb/src/mattermost/plugins/apps/server/api/impl/aws/s3.go:37 +0xdc
github.com/mattermost/mattermost-plugin-apps/server/api/impl/aws.(*Client).provisionAssets(0xc00025d540, 0xc00057c000, 0xc00025ce80, 0x2, 0x2, 0xc00025cec0, 0x0)
	/home/sumacheb/src/mattermost/plugins/apps/server/api/impl/aws/provision_app.go:305 +0x216
github.com/mattermost/mattermost-plugin-apps/server/api/impl/aws.(*Client).ProvisionApp(0xc00025d540, 0xc001872000, 0x13452a7, 0x1530000, 0x1530000, 0x0, 0x0)
	/home/sumacheb/src/mattermost/plugins/apps/server/api/impl/aws/provision_app.go:157 +0x12a6
github.com/mattermost/mattermost-plugin-apps/server/api/impl/aws.(*Client).ProvisionAppFromFile(0xc00025d540, 0x7ffe8f3cdf94, 0x3d, 0xc0002c5d00, 0x0, 0x0)
	/home/sumacheb/src/mattermost/plugins/apps/server/api/impl/aws/provision_app.go:76 +0x116
main.glob..func3(0x14f1000, 0xc000322db0, 0x1, 0x1, 0x0, 0x0)
	/home/sumacheb/src/mattermost/plugins/apps/cmd/appsctl/provision.go:39 +0x6b
github.com/spf13/cobra.(*Command).execute(0x14f1000, 0xc000322d90, 0x1, 0x1, 0x14f1000, 0xc000322d90)
	/home/sumacheb/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:850 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0x14f0ac0, 0x401260, 0x0, 0x0)
	/home/sumacheb/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/home/sumacheb/go/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
main.main()
	/home/sumacheb/src/mattermost/plugins/apps/cmd/appsctl/main.go:22 +0x31
```

This can be resolved by using `c.Service()` in https://github.com/mattermost/mattermost-plugin-apps/blob/09c6875765179df488c1675ab8035c457dfde73f/server/api/impl/aws/s3.go#L34 instead `c.service`. Would you mind also fixing that in your PR?

hanzei commented on server/api/impl/aws/provision_data.go:
```diff
@@ -111,9 +111,13 @@ func getProvisionData(b []byte, logger log) (*ProvisionData, error) {
 				File: assetFile,
 			})
-			logger.Debug("Found function bundle", "file", file.Name)
+			if logger != nil {
```
Can that case ever happen?

iomodo replied:
Yes, function `aws.GetProvisionDataFromFile` could be called with `nil`

hanzei reviewed (approved):
LGTM :+1:

levb commented:
@iomodo: I missed it on the first pass, but committing a 19Mb binary test file is not good, bloats the repo size. Can we maybe commit a dummy archive? It does not need to be a 20Mb file.

I rebasing this PR with `master`, WIP.

iomodo commented:
Additional changes:
1. rebased with master
2. renamed awsclient package to aws
3. added `provision_app` and `provision_data` in aws package
4. added fake but small app bundle, updated unit tests
cc @levb

levb reviewed (approved):
Nice job, @iomodo!

A couple follow ups to consider for later:
- consider moving `apps/aws.go` also to `aws`
- consider naming the interface method `ProvisionApp` and renaming secondary helper/impl methods as needed t avoid conflicts

LGTM!
